File: Esercizio_2/srcjava/editdistance/EditDistance.java
package editdistance;

public class EditDistance {
/**
 * @param s1: a string
 * @param s2: a string
 * @return the edit distance between two strings
 */
  public int edit_distance(String s1, String s2) {
    if (s1.length() == 0)
      return s2.length();
    
    if (s2.length() == 0)
      return s1.length();
    
    if (s1.charAt(0) == s2.charAt(0))
      return edit_distance(s1.substring(1,s1.length()),s2.substring(1,s2.length()));
    
    return 1 + Math.min(edit_distance(s1.substring(0,s1.length()),s2.substring(1,s2.length())) , 
      edit_distance(s1.substring(1,s1.length()),s2.substring(0,s2.length())));
  }
/**
 * The method is recursive with the optimization of the dynamic programming.
 * @param s1: a string
 * @param s2: a string
 * @return  the edit distance between two strings 
 */
  public int edit_distance_dyn(String s1, String s2) {
    int[][] t = new int[s1.length() + 1][s2.length() + 1];  // matrix initialization to save the value
                                                            // of the edit distance.
    for(int i = 0; i <= s1.length(); i++)                  
      t[i][0] = i;
    for(int j = 0; j <= s2.length(); j++)
      t[0][j] = j;
    for(int i = 1; i <= s1.length(); i++){
      for(int j = 1; j <= s2.length(); j++)
        t[i][j] = -1;
    }
    return rec_edit_distance_dyn(s1,s2,s1.length(),s2.length(),t);
  }

  private int rec_edit_distance_dyn(String s1, String s2, int n, int m, int[][] t) {
    if ( n == -1)
      return m;
    if ( m == -1 )
      return n;
    if(t[n][m] != -1)
      return t[n][m];
    if(s1.charAt(n-1) == s2.charAt(m-1)) {
      if(t[n-1][m-1]!= -1)
        return t[n-1][m-1];
      else 
        return rec_edit_distance_dyn(s1,s2,n-1,m-1,t);
    }
    else {
      int m1,m2;
      
      if(t[n-1][m] != -1)   
        m1 = t[n-1][m];                
      else
        m1 = rec_edit_distance_dyn(s1, s2, n-1, m, t);      
                 
      if(t[n][m-1] != -1)               
        m2 = t[n][m-1];            
      else  
        m2 = rec_edit_distance_dyn(s1, s2, n, m-1, t);
      return t[n][m] = 1 + Math.min(m1,m2);         
    }
  }
}

File: Esercizio_2/src/edit_distance.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include "edit_distance.h"

#define MAX_BUFF 1024

static int rec_edit_distance_dyn(char *s1, char *s2, int l1, int l2, int **matrix);
static int min(int a, int b);
static int **initialize_matrix(int rows, int columns);
static void free_memory(int **matrix, int size);

int edit_distance(char *s1, char *s2)
{
  if (s1 == NULL || s2 == NULL)
    return -1;
  if (strlen(s1) == 0)
    return (int)strlen(s2);
  if (strlen(s2) == 0)
    return (int)strlen(s1);

  char *sub1 = (char *)malloc(sizeof(char) * MAX_BUFF);
  char *sub2 = (char *)malloc(sizeof(char) * MAX_BUFF);
  strcpy(sub1, s1 + 1);
  strcpy(sub2, s2 + 1);

  if (s1[0] == s2[0])
    return edit_distance(sub1, sub2);
  return 1 + min(edit_distance(s1, sub2), edit_distance(sub1, s2));
} 

int edit_distance_dyn(char *s1, char *s2)
{
  if (s1 == NULL || s2 == NULL)
    return -1;
  int l1 = (int)strlen(s1);
  int l2 = (int)strlen(s2);
  int **matrix = initialize_matrix(l1, l2);

  int ret = rec_edit_distance_dyn(s1, s2, l1, l2, matrix);
  free_memory(matrix, l1);
  return ret;
}

/*
 ** Auxiliar function of edit_distance_dyn
 ** 
 ** @param s1: First string
 ** @param s2: Second string
 ** @param l1: Length of the first string
 ** @param l2: Length of the second string 
 ** @param matrix: Matrix for the memoization of the yet calculated values 
 ** @returns int: The edit distance between the two strings passed as parameters
*/
static int rec_edit_distance_dyn(char *s1, char *s2, int l1, int l2, int **matrix)
{
  if (l1 == 0)
    return l2;
  if (l2 == 0)
    return l1;

  //if two characters are not case_sensitive equals
  if (tolower(s1[l1 - 1]) == tolower(s2[l2 - 1]))
  {
    matrix[l1][l2] = rec_edit_distance_dyn(s1, s2, l1 - 1, l2 - 1, matrix);
  }
  //if two characters are not equals
  else
  {
    if (matrix[l1 - 1][l2] == -1)
      matrix[l1 - 1][l2] = rec_edit_distance_dyn(s1, s2, l1 - 1, l2, matrix);
    if (matrix[l1][l2 - 1] == -1)
      matrix[l1][l2 - 1] = rec_edit_distance_dyn(s1, s2, l1, l2 - 1, matrix);

    return matrix[l1][l2] = 1 + min(matrix[l1 - 1][l2], matrix[l1][l2 - 1]);
  }

  return matrix[l1][l2];
}

/*
 ** Method returns lower value
 ** 
 ** @param a 
 ** @param b 
 ** @returns int: The minimum between a and b
*/
static int min(int a, int b)
{
  return a <= b ? a : b;
}

/*
 ** Method initialize the matrix with -1 values
 ** 
 ** @param rows 
 ** @param columns 
 ** @returns int **: A matrix rows x columns initialized with -1 values
*/
static int **initialize_matrix(int rows, int columns)
{
  int **matrix = (int **)malloc(sizeof(int *) * (unsigned long int)(rows + 1));

  for (int i = 0; i <= rows; i++)
    matrix[i] = (int *)malloc(sizeof(int) * (unsigned long int)(columns + 1));
  for (int i = 0; i <= rows; i++)
  {
    for (int j = 0; j <= columns; j++)
      matrix[i][j] = -1;
  }
  return matrix;
}

/*
 ** Method frees the memory occupied by the matrix
 ** 
 ** @param matrix 
 ** @param size 
*/
static void free_memory(int **matrix, int size)
{
  for (int i = 0; i <= size; i++)
    free(matrix[i]);
  free(matrix);
}